Add support for Admin Partitions (#729)
hashicorp/consul-k8s: CHANGELOG.md

* CLI
  * The `consul-k8s` CLI enables users to deploy and operate Consul on Kubernetes.
    * Support `consul-k8s install` command. [[GH-713](https://github.com/hashicorp/consul-k8s/pull/713)]
* Helm Chart
  * Add support for Admin Partitions. **(Consul Enterprise only)**
  **ALPHA** [[GH-729](https://github.com/hashicorp/consul-k8s/pull/729)]

> This feature allows Consul to be deployed across multiple Kubernetes clusters while sharing a single set of Consul

Reviewer: I think this section needs to be intended so that it all looks like a part of the admin partitions bullet point.

> servers. The services on each cluster can be independently managed. This feature in the alpha phase requires a flat pod

Reviewer: might be nice to put these requirements into a bulleted list so that it reads better.

> and node network in order for inter-partition networking to work. This feature requires TLS to be enabled. This feature

Reviewer: why is TLS required?

Author: We require HTTPS for communication to externalServers. We had that already as a requirement and I decided not to loosen it for AP.

also requires Consul Namespaces enabled. Transparent Proxy is unsupported for cross partition communication.

To enable Admin Partitions on the server cluster use the following config.

```yaml
global:
  enableConsulNamespaces: true
  tls:
    enabled: true
  image: hashicorp/consul-enterprise:1.11.0-ent-alpha

>   imageK8S: hashicorp/consul-k8s-control-plane:0.34.0

Reviewer: this image is probably not needed since it's the default, right?

Author: Good point!! 😄

  adminPartitions:
    enabled: true
server:
  exposeGossipAndRPCPorts: true
  enterpriseLicense:
    secretName: license
    secretKey: key
connectInject:
  enabled: true
  transparentProxy:
    defaultEnabled: false
  consulNamespaces:
    mirroringK8S: true
controller:
  enabled: true
```

Identify the LoadBalancer External IP of the `partition-service`

```bash
kubectl get svc consul-consul-partition-service -o json | jq -r '.status.loadBalancer.ingress[0].ip'
```

> Migrate the TLS certs from the server cluster to the workload clusters

Reviewer: ```suggestion
Migrate the TLS CA credentials from the server cluster to the workload clusters
```

```bash
kubectl get secret consul-consul-ca-key --context "server-context" -o yaml | kubectl apply --context "workload-context" -f -
kubectl get secret consul-consul-ca-cert --context "server-context" -o yaml | kubectl apply --context "workload-context" -f -
```

> Configure the Workload cluster using the following config.

Reviewer: ```suggestion
Configure the workload cluster using the following config.
```

```yaml
global:
  enabled: false
  enableConsulNamespaces: true
  image: hashicorp/consul-enterprise:1.11.0-ent-alpha
  imageK8S: hashicorp/consul-k8s-control-plane:0.34.0
  adminPartitions:
    enabled: true
    name: "alpha" # Name of Admin Partition
  tls:
    enabled: true
    caCert:
      secretName: consul-consul-ca-cert
      secretKey: tls.crt
    caKey:
      secretName: consul-consul-ca-key
      secretKey: tls.key
server:
  enterpriseLicense:
    secretName: license
    secretKey: key
externalServers:
  enabled: true
  hosts: [ "loadbalancer IP" ] # external IP of partition service LB
  tlsServerName: server.dc1.consul
client:
  enabled: true
  exposeGossipPorts: true
  join: [ "loadbalancer IP" ] # external IP of partition service LB
connectInject:
  enabled: true
  consulNamespaces:
    mirroringK8S: true
controller:
  enabled: true
```

This should lead to the workload cluster having only Consul agents that connect with the Consul server. Services in this

> cluster behave like independent services. They can be configured to communicated with services in other partitions by

Reviewer: ```suggestion
cluster behave like independent services. They can be configured to communicate with services in other partitions by
```
